OpenRA.Mods.Common/UpdateRules/Rules/20210321/UnhardcodeBaseBuilderBotModule.cs

```diff
@@ -11,27 +11,85 @@
 
 using System.Collections.Generic;
 using System.Linq;
-using OpenRA.Mods.Common.Traits;
 
 namespace OpenRA.Mods.Common.UpdateRules.Rules
 {
 	public class UnhardcodeBaseBuilderBotModule : UpdateRule
 	{
+		MiniYamlNode defences;
+
+		// excludes AttackBomber and AttackTDGunboatTurreted
+		readonly string[] attackBase = { "AttackLeap", "AttackPopupTurreted", "AttackAircraft", "AttackTesla", "AttackCharges", "AttackFollow", "AttackTurreted", "AttackFrontal", "AttackGarrisoned", "AttackOmni", "AttackSwallow" };
+		readonly string[] buildings = { "Building", "EnergyWall", "D2kBuilding" };
+
+		bool anyAdded;
+
 		public override string Name => "BaseBuilderBotModule got new fields to configure buildings that are defenses.";
 
 		public override string Description => "DefenseTypes were added.";
 
-		public override IEnumerable<string> UpdateActorNode(ModData modData, MiniYamlNode actorNode)
+		public override IEnumerable<string> BeforeUpdateActors(ModData modData, List<MiniYamlNode> resolvedActors)
+		{
+			var defences = new List<string>();
+
+			foreach (var actor in resolvedActors)
+			{
+				if (actor.Key.StartsWith('^'))
+					continue;
+
+				var isBuildable = false;
+				var isBuilding = false;
+				var canAttack = false;
+
+				foreach (var trait in actor.Value.Nodes)
+				{
+					if (trait.IsRemoval())
+						continue;
+
+					if (trait.KeyMatches("Buildable"))
+					{
+						isBuildable = true;
+						continue;
+					}
+
+					if (buildings.Any(v => trait.KeyMatches(v)))
+					{
+						isBuilding = true;
+						continue;
+					}
+
+					if (attackBase.Any(ab => trait.KeyMatches(ab)))
+						canAttack = true;
+				}
+
+				if (isBuildable && isBuilding && canAttack)
+				{
+					var name = actor.Key.ToLower();
+					if (!defences.Contains(name))
+						defences.Add(name);
+				}
+			}
+
+			this.defences = new MiniYamlNode("DefenseTypes", FieldSaver.FormatValue(defences));
+
+			yield break;
+		}
+
+		public override IEnumerable<string> AfterUpdate(ModData modData)
 		{
-			var addNodes = new List<MiniYamlNode>();
+			if (anyAdded)
+				yield return "`BaseBuilderBotModule` was unhardcoded and a new field added: `DefenseTypes`. Please verify the automated changes.";
 
-			var defense = modData.DefaultRules.Actors.Values.Where(a => a.HasTraitInfo<BuildingInfo>() && a.HasTraitInfo<AttackBaseInfo>()).Select(a => a.Name);
-			var defensetypes = new MiniYamlNode("DefenseTypes", FieldSaver.FormatValue(defense.ToList()));
-			addNodes.Add(defensetypes);
+			anyAdded = false;
+		}
 
-			foreach (var baseBuilderManager in actorNode.ChildrenMatching("BaseBuilderBotModule"))
-				foreach (var addNode in addNodes)
-					baseBuilderManager.AddNode(addNode);
+		public override IEnumerable<string> UpdateActorNode(ModData modData, MiniYamlNode actorNode)
+		{
+			foreach (var squadManager in actorNode.ChildrenMatching("BaseBuilderBotModule"))
+			{
+				squadManager.AddNode(defences);
+				anyAdded = true;
```

Comment:
Can we check if the node already exists before adding it? That way things don't go awry when the rule is run multiple times (we should do this in general).

Reply:
We never account for this in other rules

Reply:
We have in the past. Rarely. It is something we should be doing more consistently though.

Reply:
Why does it not make sense? We should do this where ever possible. There is no reason not to and it makes life easier in case mistakes happen.

Reply:
I still disagree. In many cases it may be difficult to do so and I don't think it's worth the effort